src/instrument.rs

```diff
@@ -0,0 +1,941 @@
+                Some(wasm_encoder::ExportSection::new()),
+            ),
+            InstrumentationStrategy::HostCalls => (None, None),
+        };
+
+        // The new code section, containing the instrumented code.
+        let mut new_code_section = Some(wasm_encoder::CodeSection::new());
+
+        // The number of defined functions in this Wasm module.
+        let mut defined_func_count = 0;
+        // The defined function index we are currently processing.
+        let mut current_defined_func = 0;
+
+        // The number of parameters each type defines.
+        let mut num_params_by_type = vec![];
+
+        // The type index of each defined function.
+        let mut defined_func_types = vec![];
+
+        // The number of indirect call sites we've found.
+        let mut num_indirect_call_sites = 0;
+
+        // Helper to ensure we've added our modified type section when using the
+        // host-calls strategy.
+        let mut added_precise_host_types = false;
+        let mut ensure_precise_host_type_section =
+            |new_sections: &mut Vec<_>,
+             types: Option<wasmparser::TypeSectionReader>,
+             host_call_type_index: &mut Option<u32>|
+             -> Result<()> {
+                if added_precise_host_types {
+                    return Ok(());
+                }
+                added_precise_host_types = true;
+                owned(
+                    new_sections,
+                    precise_host_calls_new_type_section(types, host_call_type_index)?,
+                );
+                Ok(())
+            };
+
+        // Helper to ensure we've added our modified import section when using
+        // the host-calls strategy.
+        let mut added_precise_host_imports = false;
+        let mut ensure_precise_host_imports_section =
+            |new_sections: &mut Vec<_>,
+             imports: Option<wasmparser::ImportSectionReader>,
+             host_call_type_index: u32,
+             host_call_func_index: &mut Option<u32>|
+             -> Result<()> {
+                if added_precise_host_imports {
+                    return Ok(());
+                }
+                added_precise_host_imports = true;
+                owned(
+                    new_sections,
+                    precise_host_calls_new_import_section(
+                        imports,
+                        host_call_type_index,
+                        host_call_func_index,
+                    )?,
+                );
+                Ok(())
+            };
+
+        loop {
+            let (consumed, payload) = match parser.parse(wasm, /* eof = */ true)? {
+                Chunk::NeedMoreData(_) => unreachable!(),
+                Chunk::Parsed { consumed, payload } => (consumed, payload),
+            };
+
+            let mut sub_validator = validator.payload(&payload)?;
+            match payload {
+                Payload::Version { .. } => {}
+
+                Payload::CustomSection(custom) => {
```

Comment:
If this is a name section (`custom.name() == "name"`), don't we need to shift function indices down for the hostcall strategy?

Reply:
Not required for semantic correctness, but would definitely be nice for debugging.

Reply:
https://github.com/fitzgen/winliner/issues/8